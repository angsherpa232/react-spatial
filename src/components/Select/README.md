#

This demonstrates the use of Select.

```jsx
import React from 'react';
import Select from 'react-spatial/components/Select';

class SelectExample extends React.Component {
  constructor(props) {
    super(props);
    this.cities = [
      { label: 'New york', value: 'ny' },
      { label: 'Los angeles', value: 'ls' },
      { label: 'Orlando', value: 'orl' },
    ];
    this.state = { city: this.cities[1] };
  }

  render() {
    return (
      <div className="tm-select-example">
        <Select
          options={this.cities}
          value={this.state.city}
          onChange={(e, city) => {
            this.setState({ city });
          }}
        />
        <div>You selected "{this.state.city.label}".</div>
      </div>
    );
  }
}

<SelectExample />;
```
