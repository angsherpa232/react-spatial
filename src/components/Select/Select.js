import React from 'react';
import PropTypes from 'prop-types';

const propTypes = {
  /**
   * CSS class of the select.
   */
  className: PropTypes.string,

  /**
   * Array of values to use in the select element.
   */
  options: PropTypes.arrayOf(
    PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
      PropTypes.shape({
        value: PropTypes.oneOfType([
          PropTypes.string,
          PropTypes.number,
          PropTypes.arrayOf(
            PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
          ),
        ]).isRequired,
        label: PropTypes.string.isRequired,
      }),
    ]),
  ),

  /**
   * Selected option object.
   */
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.shape({
      value: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
        PropTypes.arrayOf(
          PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        ),
      ]).isRequired,
      label: PropTypes.string.isRequired,
    }),
  ]),

  /**
   * Function triggered on select's change event.
   * @param {Event} event The change event object.
   */
  onChange: PropTypes.func.isRequired,
};

const defaultProps = {
  value: undefined,
  options: undefined,
  className: 'tm-select',
};

const getSelectedOption = (inputValue, options) => {
  return options.find(
    opt =>
      (typeof opt === 'string' && opt === inputValue) ||
      opt.value === inputValue ||
      (opt.value && opt.value.toString && opt.value.toString() === inputValue),
  );
};

const getValue = opt => {
  return typeof opt === 'string' ? opt : opt.value;
};

const getLabel = opt => {
  return typeof opt === 'string' ? opt : opt.label;
};

/**
 * This component displays a simple select HTML element.
 */
const Select = props => {
  const { value, className, options, onChange } = props;
  if (!options) {
    return null;
  }
  const inputValue = getValue(value);

  return (
    <select
      className={className}
      value={inputValue}
      onChange={evt => {
        onChange(evt, getSelectedOption(evt.target.value, options));
      }}
    >
      {(options || []).map(o => (
        <option key={getValue(o)} value={getValue(o)}>
          {getLabel(o)}
        </option>
      ))}
    </select>
  );
};

Select.propTypes = propTypes;
Select.defaultProps = defaultProps;

export default Select;
