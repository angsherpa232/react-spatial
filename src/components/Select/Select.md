#

This demonstrates the use of Select.

```jsx
const React = require('react');
const Select = require('./Select').default;

class SelectExample extends React.Component {
  constructor(props) {
    super(props);
    this.state = { value: 'ny' };
  }

  render() {
    return (
      <div className="tm-select-example">
        <Select
          options={[
            { label: 'New york', value: 'ny' },
            { label: 'Los angeles', value: 'ls' },
            { label: 'Orlando', value: 'orl' },
          ]}
          value={this.state.value}
          onChange={(e, value) => {
            this.setState({ value: value });
          }}
        />
        <div>You selected "{this.state.value}".</div>
      </div>
    );
  }
}

<SelectExample />;
```
