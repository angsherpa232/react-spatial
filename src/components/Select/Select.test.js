import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import Select from './Select';

configure({ adapter: new Adapter() });

describe('Select', () => {
  describe('when no properties are set', () => {
    let spy = null;

    beforeEach(() => {
      window.console.error = jest.fn().mockImplementation(() => {});
      spy = jest.spyOn(window.console, 'error');
    });

    afterEach(() => {
      spy.mockRestore();
      window.console.error.mockRestore();
    });

    test('matches snapshot', () => {
      const component = renderer.create(<Select />);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('when options are strings', () => {
    const options = ['foo', 'bar', 'baz'];

    const onChange = () => {};

    test('matches snapshot', () => {
      const component = renderer.create(
        <Select options={options} value="bar" onChange={() => {}} />,
      );

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('executes onChange property function', () => {
      const onChangeMock = jest.fn(onChange);
      const wrapper = shallow(
        <Select options={options} value="baz" onChange={onChangeMock} />,
      );
      expect(onChangeMock).toHaveBeenCalledTimes(0);
      const evt = { target: { value: 'foo' } };
      wrapper.find('select').simulate('change', evt);
      expect(onChangeMock).toHaveBeenCalledTimes(1);
      expect(onChangeMock).toHaveBeenCalledWith(evt, options[0]);
    });
  });

  describe('when options are objects with a value property.', () => {
    const options = [
      {
        value: 'foo',
        label: 'qux',
      },
      {
        value: 'bar',
        label: 'quux',
      },
      {
        value: 'baz',
        label: 'corge',
      },
    ];

    const onChange = () => {};

    test('matches snapshot', () => {
      const component = renderer.create(
        <Select options={options} value={options[1]} onChange={() => {}} />,
      );

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('executes onChange property function', () => {
      const onChangeMock = jest.fn(onChange);
      const wrapper = shallow(
        <Select options={options} value={options[2]} onChange={onChangeMock} />,
      );
      expect(onChangeMock).toHaveBeenCalledTimes(0);
      const evt = { target: { value: 'foo' } };
      wrapper.find('select').simulate('change', evt);
      expect(onChangeMock).toHaveBeenCalledTimes(1);
      expect(onChangeMock).toHaveBeenCalledWith(evt, options[0]);
    });
  });
});
