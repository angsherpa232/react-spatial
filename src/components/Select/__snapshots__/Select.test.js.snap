// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select when no properties are set matches snapshot 1`] = `null`;

exports[`Select when options are objects with a value property. matches snapshot 1`] = `
<select
  className="tm-select"
  onChange={[Function]}
  value="bar"
>
  <option
    value="foo"
  >
    qux
  </option>
  <option
    value="bar"
  >
    quux
  </option>
  <option
    value="baz"
  >
    corge
  </option>
</select>
`;

exports[`Select when options are strings matches snapshot 1`] = `
<select
  className="tm-select"
  onChange={[Function]}
  value="bar"
>
  <option
    value="foo"
  >
    foo
  </option>
  <option
    value="bar"
  >
    bar
  </option>
  <option
    value="baz"
  >
    baz
  </option>
</select>
`;
